src/xenia/hid/winkey/winkey_input_driver.cc

```diff
@@ -10,17 +10,121 @@
 #include "xenia/hid/winkey/winkey_input_driver.h"
 
 #include "xenia/base/platform_win.h"
+#include "xenia/base/logging.h"
 #include "xenia/hid/hid_flags.h"
 #include "xenia/hid/input_system.h"
 #include "xenia/ui/virtual_key.h"
 #include "xenia/ui/window.h"
 
+DEFINE_string(keybind_dpad_left,         "^'A'",    "XINPUT_GAMEPAD_DPAD_LEFT"        , "KeyBind")
+DEFINE_string(keybind_dpad_right,        "^'D'",    "XINPUT_GAMEPAD_DPAD_RIGHT"       , "KeyBind")
+DEFINE_string(keybind_dpad_down,         "^'S'",    "XINPUT_GAMEPAD_DPAD_DOWN"        , "KeyBind")
+DEFINE_string(keybind_dpad_up,           "^'W'",    "XINPUT_GAMEPAD_DPAD_UP"          , "KeyBind")
+DEFINE_string(keybind_left_thumb_left,   "_'A'",    "XINPUT_GAMEPAD_LEFT_THUMB_LEFT"  , "KeyBind")
+DEFINE_string(keybind_left_thumb_right,  "_'D'",    "XINPUT_GAMEPAD_LEFT_THUMB_RIGHT" , "KeyBind")
+DEFINE_string(keybind_left_thumb_down,   "_'S'",    "XINPUT_GAMEPAD_LEFT_THUMB_DOWN"  , "KeyBind")
+DEFINE_string(keybind_left_thumb_up,     "_'W'",    "XINPUT_GAMEPAD_LEFT_THUMB_UP"    , "KeyBind")
+DEFINE_string(keybind_left_thumb,        "'F'",     "XINPUT_GAMEPAD_LEFT_THUMB"       , "KeyBind")
+DEFINE_string(keybind_right_thumb_up,    "0x26",    "XINPUT_GAMEPAD_RIGHT_THUMB_UP"   , "KeyBind")
+DEFINE_string(keybind_right_thumb_down,  "0x28",    "XINPUT_GAMEPAD_RIGHT_THUMB_DOWN" , "KeyBind")
+DEFINE_string(keybind_right_thumb_right, "0x27",    "XINPUT_GAMEPAD_RIGHT_THUMB_RIGHT", "KeyBind")
+DEFINE_string(keybind_right_thumb_left,  "0x25",    "XINPUT_GAMEPAD_RIGHT_THUMB_LEFT" , "KeyBind")
+DEFINE_string(keybind_right_thumb,       "'K'",     "XINPUT_GAMEPAD_RIGHT_THUMB"      , "KeyBind")
+DEFINE_string(keybind_x,                 "'L'",     "XINPUT_GAMEPAD_X"                , "KeyBind")
+DEFINE_string(keybind_b,                 "0xDE",    "XINPUT_GAMEPAD_B"                , "KeyBind")
+DEFINE_string(keybind_a,                 "0xBA",    "XINPUT_GAMEPAD_A"                , "KeyBind")
+DEFINE_string(keybind_y,                 "'P'",     "XINPUT_GAMEPAD_Y"                , "KeyBind")
+DEFINE_string(keybind_left_trigger,      "'Q' 'I'", "XINPUT_GAMEPAD_LEFT_TRIGGER"     , "KeyBind")
+DEFINE_string(keybind_right_trigger,     "'E' 'O'", "XINPUT_GAMEPAD_RIGHT_TRIGGER"    , "KeyBind")
+DEFINE_string(keybind_back,              "'Z'",     "XINPUT_GAMEPAD_BACK"             , "KeyBind")
+DEFINE_string(keybind_start,             "'X'",     "XINPUT_GAMEPAD_START"            , "KeyBind")
+DEFINE_string(keybind_left_shoulder,     "'1'",     "XINPUT_GAMEPAD_LEFT_SHOULDER"    , "KeyBind")
+DEFINE_string(keybind_right_shoulder,    "'3'",     "XINPUT_GAMEPAD_RIGHT_SHOULDER"   , "KeyBind")
+
 namespace xe {
 namespace hid {
 namespace winkey {
 
+void WinKeyInputDriver::ParseKeyBindingCVar(ui::VirtualKey controllerKey, std::string keyBindCVar) {
+  const char *cstr = keyBindCVar.c_str();
+  char token[16];
+  std::size_t off = 0;
+  while (keyBindCVar[off] != 0) {
```

Comment:
`std::string` is not obligated to be null terminated, this should be a length check.